File: C#/Leetcode/Array/LongestIncreasingSubsequence.cs
using System;

namespace LeetcodeSolutions.Array
{
    // Leetcode 300
    // Dynamic Programming - Common subproblems
    // Reference - Explanation: https://www.youtube.com/watch?v=CE2b_-XfVDk
    //           - Code : https://github.com/mission-peace/interview/blob/master/src/com/interview/dynamic/LongestIncreasingSubsequence.java
    // Submission - https://leetcode.com/submissions/detail/140963157/
    public class LongestIncreasingSubsequence
    {
        //public static void Main(string[] args)
        //{
        //    LongestIncreasingSubsequence LIS = new LongestIncreasingSubsequence();

        //    int[] arr1 = { 3, 4, -1, 0, 6, 2, 3 }; // 4
        //    Console.WriteLine(LIS.LengthOfLIS(arr1));

        //    int[] arr2 = { 3, 4, 5, 6, 7, 8, 9 }; // 7
        //    Console.WriteLine(LIS.LengthOfLIS(arr2));

        //    int[] arr3 = { 1, 3, -1, 0, 2, -2 }; // 4
        //    Console.WriteLine(LIS.LengthOfLIS(arr3));

        //    int[] arr4 = { 6, 5, 4 }; // 1
        //    Console.WriteLine(LIS.LengthOfLIS(arr4));

        //    int[] arr5 = { }; // 0
        //    Console.WriteLine(LIS.LengthOfLIS(arr5));

        //    Console.ReadKey();
        //}

        // TODO: There is a O(nlogn) solution.
        // Tx = O(n^2)
        // Sx = O(n)
        public int LengthOfLIS(int[] nums)
        {
            int length = nums.Length;

            int[] LIS = new int[length];

            // Each number has a sequence of 1.
            for (int i = 0; i < length; i++)
                LIS[i] = 1;

            // Initialize to 0 to cover the case of array is empty
            int maxLIS = length == 0 ? 0 : 1;

            for (int i = 1; i < length; i++)
            {
                for (int j = 0; j < i; j++)
                {
                    if (nums[j] < nums[i])
                    {
                        // Max of the previously computed LIS of J and current number LIS.
                        LIS[i] = System.Math.Max(LIS[i], LIS[j] + 1);
                        maxLIS = System.Math.Max(maxLIS, LIS[i]);
                    }
                }
            }
            
            return maxLIS;
        }
    }
}

File: Java/DynamicProgramming/LongestIncreasingSubsequence.java
package DynamicProgramming;

public class LongestIncreasingSubsequence {
    // Leetcode 300: https://leetcode.com/problems/longest-increasing-subsequence/description/
    // Submission Detail: https://leetcode.com/problems/longest-increasing-subsequence/submissions/1662367078

    //Input: nums = [8,2,4,3,6,12]
    //Output:  4

    //Input: nums = [10,9,2,5,3,7,101,18]
    //Output: 4
    //Explanation: The longest increasing subsequence is [2,3,7,101], therefore the length is 4.

    //Input: nums = [0,1,0,3,2,3]
    //Output: 4

    //Input: nums = [7,7,7,7,7,7,7]
    //Output: 1

    //Input: nums = [7,6,5,4,3,2,1]
    //Output: 1

    // Tx = O(n*n)
    // Sx = O(n)
    public int lengthOfLIS(int[] nums) {
        int[] dp = new int[nums.length];
        int maxLength = 1;  // initialize to 1 because the nums could be in decreasing order. And max is atleast 1.

        // Base Case: Each value in the nums has a minimum length of 1
        for(int i=0; i<dp.length; i++)
            dp[i] = 1;

        for(int i=0; i<nums.length; i++) {
            for(int j=0; j<i; j++) {
                if(nums[i] > nums[j]) {
                    dp[i] = Math.max(dp[i], dp[j]+1); // +1 because current number at i is greater than j in nums.
                    maxLength = Math.max(maxLength, dp[i]);
                }
            }
        }

        return maxLength;
    }
}
